src/core/support/alloc.c

```diff
@@ -34,13 +34,33 @@
 #include <grpc/support/alloc.h>
 
 #include <stdlib.h>
+#include <grpc/support/log.h>
 #include <grpc/support/port_platform.h>
 #include "src/core/profiling/timers.h"
 
+static void *(*malloc_fn)(size_t size) = malloc;
+static void *(*realloc_fn)(void *ptr, size_t size) = realloc;
+static void (*free_fn)(void *ptr) = free;
+
+void gpr_set_malloc(void *(*custom_malloc)(size_t size)) {
```

Comment:
Interface tweak to prevent mis-setting these functions:

typedef struct {
  void (*malloc_fn)(size_t);
  /* ... */
} gpr_allocation_functions;

void gpr_set_allocation_functions(gpr_allocation_functions functions) {
  GPR_ASSERT(functions.malloc_fn != NULL);
  GPR_ASSERT(...);
  g_functions = functions;
}

This should be exposed in support/alloc.h

And of course this needs a test

(alloc_test.c?)

Documentation needs to be clear that this is best effort and we do not guarantee that NO malloc/free() calls will be made.